---
output: html_document
---
This lab on the Introduction to R comes from "Introduction to Statistical Learning with Applications in R" by Gareth James, Daniela Witten, Trevor Hastie and Robert Tibshirani. It was re-implemented in Fall 2016 in `tidyverse` format by Amelia McNamara and R. Jordan Crouser at Smith College.

# The `tidyverse`

The textbook was written using `base` `R` functions, which can be unintuitive. For our labs, we will be using the so-called 'tidyverse,' which is based on the idea of tidy data and functions. Tidy data has observations as rows and variables as columns. All the observations must be at the same observational level. Tidy functions are pure, pipeable, and predictable. We'll begin to understand all of this more fully as we start to work with data.

Because the tidyverse is outside the `base` `R` functions, we must use packages (sometimes called libraries) to get the additional functionality. Packages are discussed further in Chapter 3. For now, lets just load a bunch of packages we'll need for our work.

In [1]:
library(ISLR)
library(dplyr)
library(readr)
library(ggplot2)
library(GGally)
library(mosaic)
library(manipulate)


Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union



ERROR: Error in library(ggplot2): there is no package called ‘ggplot2’


Getting error messages? You may need to install the packages first using `install.packages()`.

# Loading Data

For most analyses, the first step involves importing a data set into `R`. For this class, a lot of the data comes from the package itself, so we can use the `data()` function to import it.

We begin by loading in the `Auto` data set. This data is part of the `ISLR` package.

In [2]:
data(Auto)

Nothing happens when you run this, but now the data is available in your environment. (In RStudio, you would see the name of the data in your Environment tab).

To view the data, we can either print the entire dataset by typing its name, or we can "slice" some of the data off to look at just a subset by piping data using the `%>%` operator into the `slice` function:

In [3]:
Auto
Auto %>% 
    slice(1:10)

mpg cylinders displacement horsepower weight acceleration year origin
1   18  8         307          130        3504   12.0         70   1     
2   15  8         350          165        3693   11.5         70   1     
3   18  8         318          150        3436   11.0         70   1     
4   16  8         304          150        3433   12.0         70   1     
5   17  8         302          140        3449   10.5         70   1     
6   15  8         429          198        4341   10.0         70   1     
7   14  8         454          220        4354    9.0         70   1     
8   14  8         440          215        4312    8.5         70   1     
9   14  8         455          225        4425   10.0         70   1     
10  15  8         390          190        3850    8.5         70   1     
11  15  8         383          170        3563   10.0         70   1     
12  14  8         340          160        3609    8.0         70   1     
13  15  8         400          150        3761    9.5         70   1     
14  14  8         455          225        3086   10.0         70   1     
15  24  4         113           95        2372   15.0         70   3     
16  22  6         198           95        2833   15.5         70   1     
17  18  6         199           97        2774   15.5         70   1     
18  21  6         200           85        2587   16.0         70   1     
19  27  4          97           88        2130   14.5         70   3     
20  26  4          97           46        1835   20.5         70   2     
21  25  4         110           87        2672   17.5         70   2     
22  24  4         107           90        2430   14.5         70   2     
23  25  4         104           95        2375   17.5         70   2     
24  26  4         121          113        2234   12.5         70   2     
25  21  6         199           90        2648   15.0         70   1     
26  10  8         360          215        4615   14.0         70   1     
27  10  8         307          200        4376   15.0         70   1     
28  11  8         318          210        4382   13.5         70   1     
29   9  8         304          193        4732   18.5         70   1     
30  27  4          97           88        2130   14.5         71   3     
⋮   ⋮   ⋮         ⋮            ⋮          ⋮      ⋮            ⋮    ⋮     
368 28  4         112           88        2605   19.6         82   1     
369 27  4         112           88        2640   18.6         82   1     
370 34  4         112           88        2395   18.0         82   1     
371 31  4         112           85        2575   16.2         82   1     
372 29  4         135           84        2525   16.0         82   1     
373 27  4         151           90        2735   18.0         82   1     
374 24  4         140           92        2865   16.4         82   1     
375 36  4         105           74        1980   15.3         82   2     
376 37  4          91           68        2025   18.2         82   3     
377 31  4          91           68        1970   17.6         82   3     
378 38  4         105           63        2125   14.7         82   1     
379 36  4          98           70        2125   17.3         82   1     
380 36  4         120           88        2160   14.5         82   3     
381 36  4         107           75        2205   14.5         82   3     
382 34  4         108           70        2245   16.9         82   3     
383 38  4          91           67        1965   15.0         82   3     
384 32  4          91           67        1965   15.7         82   3     
385 38  4          91           67        1995   16.2         82   3     
386 25  6         181          110        2945   16.4         82   1     
387 38  6         262           85        3015   17.0         82   1     
388 26  4         156           92        2585   14.5         82   1     
389 22  6         232          112        2835   14.7         82   1     
390 32  

mpg cylinders displacement horsepower weight acceleration year origin
1  18  8         307          130        3504   12.0         70   1     
2  15  8         350          165        3693   11.5         70   1     
3  18  8         318          150        3436   11.0         70   1     
4  16  8         304          150        3433   12.0         70   1     
5  17  8         302          140        3449   10.5         70   1     
6  15  8         429          198        4341   10.0         70   1     
7  14  8         454          220        4354    9.0         70   1     
8  14  8         440          215        4312    8.5         70   1     
9  14  8         455          225        4425   10.0         70   1     
10 15  8         390          190        3850    8.5         70   1     
   name                     
1  chevrolet chevelle malibu
2  buick skylark 320        
3  plymouth satellite       
4  amc rebel sst            
5  ford torino              
6  ford galaxie 500         
7  chevrolet impala         
8  plymouth fury iii        
9  pontiac catalina         
10 amc ambassador dpl

If data isn't internal to packages, .csv files are the most common way for data to be formatted. CSV files are simply text files, which you could alternatively open on your computer using a standard text editor. While it is often tempting to view a data set using Excel before loading it into `R`, it can introduce problems in the data later. 

We can use the `read_csv()` function in the `readr` package to load in a CSV file. The help file contains details about how to use this function.

Before attempting to load a data set, we must make sure that `R` knows to search for the data in the proper directory. If you are using internal package data, this isn't a concern-- just make sure you have run `library()` on the package containing the data. For external data, the easiest way make sure the data is accessible to R is to put the data in the same folder as your Jupyter notebook. We have a file called `Auto.csv` in this directory, so we can read it in without worrying about the file path. 

When we read in data that is external, we must give it a name. We'll call it `Auto` and use the assignment operator `=`. The assignment operator `<-` is equivalent, and using one or the other is a matter of personal preference.

In [5]:
Auto=read_csv("Auto.csv", na="?")
Auto %>% 
    slice(1:10)

ERROR: Error: 'Auto.csv' does not exist in current working directory ('/home/iserina/Notebook/mldmlab').


Notice that the data looks just the same as when we loaded it from the package. Now that we have the data, we can begin to learn things about it.

In [6]:
dim(Auto)

[1] 392   9

In [7]:
str(Auto)

'data.frame':	392 obs. of  9 variables:
 $ mpg         : num  18 15 18 16 17 15 14 14 14 15 ...
 $ cylinders   : num  8 8 8 8 8 8 8 8 8 8 ...
 $ displacement: num  307 350 318 304 302 429 454 440 455 390 ...
 $ horsepower  : num  130 165 150 150 140 198 220 215 225 190 ...
 $ weight      : num  3504 3693 3436 3433 3449 ...
 $ acceleration: num  12 11.5 11 12 10.5 10 9 8.5 10 8.5 ...
 $ year        : num  70 70 70 70 70 70 70 70 70 70 ...
 $ origin      : num  1 1 1 1 1 1 1 1 1 1 ...
 $ name        : Factor w/ 304 levels "amc ambassador brougham",..: 49 36 231 14 161 141 54 223 241 2 ...


In [ ]:
names(Auto)

The `dim()` function tells us that the data has 392 observations, or rows, and
nine variables, or columns. The original data had some empty rows, but when we read the data in `R` knew to ignore them. 

The ` str()` function tells us that most of the variables are numeric or integer, although the ` name ` variable is a character vector. 

` names()` lets us check the variable names.

# Summary statistics

Often, we want to know some basic things about variables in our data. `summary()` on an entire dataset will give you an idea of some of the distributions of your variables.

The ` summary()` function produces a numerical summary of each variable in
a particular data set.

In [ ]:
 summary(Auto)

The summary suggests that `origin` might be better thought of as a factor. It only seems to have three possible values, `1`, `2` and `3`. If we read the documentation about the data (using `?Auto`) we will learn that these numbers correspond to where the car is from: 1. American, 2. European, 3. Japanese.

In [ ]:
?Auto

So, lets `mutate()` that variable into a factor (categorical) variable.

In [ ]:
Auto = Auto %>%
    mutate(origin = factor(origin))
summary(Auto)

We can also look at statistics for a particular variable with the function `favstats()`

In [ ]:
favstats(~mpg, data=Auto)

Or, just look at one particular statistic using `mean()`, `sd()`, `median()`, `IQR()`, and more.

In [ ]:
mean(~displacement, data=Auto)

# Plotting

We can use the `ggplot2` package to produce simple graphics. `ggplot2` has a particular syntax, which looks like this

In [8]:
ggplot(Auto) + geom_point(aes(x=cylinders, y=mpg))

ERROR: Error in ggplot(Auto): non trovo la funzione "ggplot"


The basic idea is that you need to initialize a plot with `ggplot()` and then add "geoms" (short for geometric objects) to the plot. The `ggplot2` package is based on the [Grammar of Graphics](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=0ahUKEwjV6I6F4ILPAhUFPT4KHTFiBwgQFggcMAA&url=https%3A%2F%2Fwww.amazon.com%2FGrammar-Graphics-Statistics-Computing%2Fdp%2F0387245448&usg=AFQjCNF5D6H3ySCsgqBTdp96KNF3bGyU2Q&sig2=GnNgoN6Ztn3AJSTJYaMPwA), a famous book on data visualization theory. It is a way to map attributes in your data (like variables) to "aesthetics" on the plot. The parameter `aes()` is short for aesthetic. 

For more about the `ggplot2` syntax, view the help by typing `?ggplot` or `?geom_point`. There are also great online resources for `ggplot2`, like the [R graphics cookbook](http://www.cookbook-r.com/Graphs/).

In [ ]:
?ggplot

The ` cylinders` variable is stored as a numeric vector, so `R` has treated it
as quantitative. However, since there are only a small number of possible
values for cylinders, one may prefer to treat it as a qualitative variable.
We can turn it into a factor, again using a `mutate()` call.

In [ ]:
Auto = Auto %>%
    mutate(cylinders = factor(cylinders))

To view the relationship between a categorical and a numeric variable, we might want to produce _boxplots_. As usual, a number of options can be specified in order to customize the plots.

In [ ]:
ggplot(Auto) + geom_boxplot(aes(x=cylinders, y=mpg)) + xlab("Cylinders") + ylab("MPG")

The geom `geom_histogram()` can be used to plot a histogram.

In [ ]:
ggplot(Auto) + geom_histogram(aes(x=mpg))

The function warns us that it used a default number of bins, so we should think more carefully about what value makes sense.

In [ ]:
ggplot(Auto) + geom_histogram(aes(x=mpg), binwidth=5) 

For small datasets, we might want to see all the bivariate relationships between the variables. The `GGally` package has an extension of the scatterplot matrix that can do just that. (Be patient-- it takes a long time!)

In [ ]:
Auto %>% ggpairs()

We can also produce a pairs plot for just a subset of the variables by piping just a selection of the variables into the plot.

In [ ]:
Auto %>% select(mpg, cylinders) %>%
    ggpairs()

Sometimes, we might want to save a plot for use outside of R. To do this, we can use the `ggsave()` function.

In [ ]:
ggsave("histogram.png",
       ggplot(Auto) + geom_histogram(aes(x=mpg), binwidth=5) 
      )

# Appendices

These sections are not required in order to understand the tidyverse material we are focusing on, but are part of the lab in the book, so we are including them here. The appendices start with making your own vectors and matrices, something which is almost never required in the tidyverse. Then, the appendices move on to more sophisticated plots (such as those in 3 dimensions) and finish up with some introduction to indexing using numbers. With our tidy tools, we will try to avoid indexing using numbers if at all possible, but you may want to look at this material to see another method for slicing data.

# Making vectors and matrices

`R` uses functions to perform operations. To run a function called `funcname`, we type `funcname(input1, input2)`, where the inputs (or arguments) `input1` and `input2` tell `R` how to run the function. A function can have any number of inputs. 

For example, to create a vector of numbers, we use the function
`c()` (for concatenate). Any numbers inside the parentheses are joined together. The following command instructs `R` to join together the numbers 1, 3, 2, and 5, and to save them as a vector named `x`. When we type `x`, it gives us back the vector.

In [ ]:
x <- c(1,3,2,5)
x

We can also save things using $=$ rather than $<-$:

In [ ]:
x = c(1,6,2)
x

Hitting the up arrow multiple times will display the previous commands,
which can then be edited. This is useful since one often wishes to repeat
a similar command. In addition, typing `?funcname` will always cause `R` to
open a new help file window with additional information about the function
`funcname`.

In [ ]:
?c

We can tell `R` to add two sets of numbers together. It will then add the
first number from `x` to the first number from `y`, and so on. To do this, `x` and
`x` should be the same length. We can check the length of each vector using the `length()`
function:

In [ ]:
y = c(1,4,3)
length(x)==length(y)

Looks good, so let's sum them up (I'll print the starting value of the `x` and `y` vectors, too):

In [ ]:
#Value of x
x

#Value of y
y

#Value of x+y
x+y

The `ls()` function allows us to look at a list of all of the objects, such as data and functions, that we have saved so far:

In [ ]:
ls()

The `rm()` function can be used to delete any that we don’t want:

In [ ]:
rm(x)
ls()

It’s also possible to remove all objects at once (effectively clearing the history):

In [ ]:
rm(list=ls())

The `matrix()` function can be used to create a matrix of numbers. Before
we use the `matrix()` function, we can learn more about it:

In [ ]:
?matrix

The help file reveals that the `matrix()` function takes a number of inputs,
but for now we focus on the first three: the data (the entries in the matrix),
the number of rows, and the number of columns. First, we create a simple
matrix.

In [ ]:
x=matrix(data=c(1,2,3,4,5,6), ncol=2, nrow=3)
x

Note that we could just as well omit typing `data=`, `nrow=`, and `ncol=` in the
`matrix()` command above: that is, we could just type

In [ ]:
x=matrix(c(1,2,3,4,5,6),3,2)
x

and this would have the same effect. However, it can sometimes be useful to
specify the names of the arguments passed in, since otherwise `R` will assume
that the function arguments are passed into the function in the same order
that is given in the function’s help file. As this example illustrates, by
default `R` creates matrices by successively filling in columns. Alternatively,
the `byrow=TRUE` option can be used to populate the matrix in order of the
rows.

In [ ]:
matrix(c(1,2,3,4),2,2,byrow=TRUE)
matrix(c(1,2,3,4),2,2,byrow=FALSE)

(Notice that in the above command we did not assign the matrix to a value
such as `x`. In this case the matrix is printed to the screen but is not saved
for future calculations.)

The `sqrt()` function returns the square root of each element of a vector or matrix. The command `x^2` raises each element of `x` to the power 2; any powers are possible, including fractional or negative
powers.

In [ ]:
# orginal value of x
x

# square root
sqrt(x)

# values squared
x^2

The `rnorm()` function generates a vector of random normal variables,
with first argument `n` the sample size. Each time we call this function, we
will get a different answer. Here we create two correlated sets of numbers,
`x` and `y`, and use the `cor()` function to compute the correlation between
them.

In [ ]:
x=rnorm(50)
y=x+rnorm(50,mean=50,sd=5)
cor(x,y)

By default, `rnorm()` creates standard normal random variables with a mean
of 0 and a standard deviation of 1. However, the mean and standard deviation
can be altered using the `mean` and `sd` arguments, as illustrated above.
Sometimes we want our code to reproduce the exact same set of random
numbers; we can use the `set.seed()` function to do this. The `set.seed()`
function takes an (arbitrary) integer argument.

In [ ]:
set.seed(13)
rnorm(50)

We'll use `set.seed()` throughout the labs whenever we perform calculations
involving random quantities. In general this should allow each of you to reproduce
the results from the book. Depending on your version of `R` it is possible that some small discrepancies may appear between the book and your output from `R`.

The `mean()` and `var()` functions can be used to compute the mean and
variance of a vector of numbers. Applying `sqrt()` to the output of `var()`
will give the standard deviation. Or we can simply use the `sd()` function.

In [ ]:
set.seed(3)
y=rnorm(100)
y[0:10]

# Compute the mean
mean(y)

# Compute the variance
var(y)

# Standard deviation = square root of variance
sqrt(var(y))

# But there's a function for that
sd(y)

# Graphics

We will now create some more sophisticated plots. The `contour()` function produces a contour plot in order to represent three-dimensional data; it is like a topographical map. It takes three arguments:

1. A vector of the `x` values (the first dimension),
2. A vector of the `y` values (the second dimension), and
3. A `matrix` whose elements correspond to the `z` value (the third dimension) for each pair of `(x,y)` coordinates.

As with the `plot()` function, there are many other inputs that can be used
to fine-tune the output of the `contour()` function. To learn more about
these, take a look at the help file by typing `?contour`.

In [ ]:
y=x
f=outer(x,y,function(x,y)cos(y)/(1+x^2))
contour(x,y,f)
contour(x,y,f,nlevels=45,add=T)
fa=(f-t(f))/2
contour(x,y,fa,nlevels=15)

The `image()` function works the same way as `contour()`, except that it
produces a color-coded plot whose colors depend on the `z` value. This is known as a heatmap, and is sometimes used to plot temperature in weather
forecasts. Alternatively, `persp()` can be used to produce a three-dimensional
plot. The arguments `theta` and `phi` control the angles at which the plot is
viewed.

In [ ]:
image(x,y,fa)
persp(x,y,fa,theta=30,phi=40)

# Indexing Data

We often wish to examine part of a set of data. Suppose that our data is
stored in the matrix `A`.

In [ ]:
A=matrix(1:16,4,4)
A

Then, typing:

In [ ]:
A[2,3]

will select the element corresponding to the second row and the third column.
The first number after the open-bracket symbol `[` always refers to
the row, and the second number always refers to the column. We can also
select multiple rows and columns at a time, by providing vectors as the
indices.

In [ ]:
A[c(1,3),c(2,4)]
A[1:3,2:4]
A[1:2,]
A[,1:2]

The last two examples include either no index for the columns or no index
for the rows. These indicate that `R` should include all columns or all rows,
respectively. `R` treats a single row or column of a matrix as a vector.

In [ ]:
A[,1]

The use of a negative sign `-` in the index tells R to keep all rows or columns
_except_ those indicated in the index.

In [ ]:
A[-c(1,3),]
A[-c(1,3),-c(1,3,4)]

The `dim()` function outputs the dimensions of a matrix, written as the number of rows followed by the number of
columns.

In [ ]:
dim(A)